File: training-data/swift/priority_queue.swift
final class PriorityQueue {
    private var heap: [Int] = []

    var isEmpty: Bool { heap.isEmpty }

    func push(_ value: Int) {
        heap.append(value)
        var i = heap.count - 1
        while i > 0 {
            let parent = (i - 1) / 2
            if heap[parent] <= heap[i] { break }
            heap.swapAt(parent, i)
            i = parent
        }
    }

    func pop() -> Int? {
        guard !heap.isEmpty else { return nil }
        let top = heap[0]
        let last = heap.removeLast()
        if !heap.isEmpty {
            heap[0] = last
            var i = 0
            while true {
                let left = 2 * i + 1
                let right = 2 * i + 2
                var smallest = i
                if left < heap.count && heap[left] < heap[smallest] { smallest = left }
                if right < heap.count && heap[right] < heap[smallest] { smallest = right }
                if smallest == i { break }
                heap.swapAt(smallest, i)
                i = smallest
            }
        }
        return top
    }
}

let pq = PriorityQueue()
for v in [5, 3, 8, 1, 9, 2] {
    pq.push(v)
}

var sorted: [Int] = []
while let v = pq.pop() {
    sorted.append(v)
}
print(sorted)

File: training-data/clojure/priority_queue.clj
(defn pq-parent [i] (quot (dec i) 2))
(defn pq-left [i] (inc (* 2 i)))
(defn pq-right [i] (+ 2 (* 2 i)))

(defn pq-push [heap value]
  (let [h0 (conj heap value)]
    (loop [h h0 i (dec (count h0))]
      (if (and (pos? i) (< (nth h i) (nth h (pq-parent i))))
        (let [p (pq-parent i)]
          (recur (assoc h i (nth h p) p (nth h i)) p))
        h))))

(defn pq-sift-down [heap i]
  (let [n (count heap)
        l (pq-left i)
        r (pq-right i)
        smallest (if (and (< l n) (< (nth heap l) (nth heap i))) l i)
        smallest (if (and (< r n) (< (nth heap r) (nth heap smallest))) r smallest)]
    (if (= smallest i)
      heap
      (recur (assoc heap i (nth heap smallest) smallest (nth heap i)) smallest))))

(defn pq-pop [heap]
  (let [n (count heap)]
    (if (<= n 1)
      [(first heap) []]
      (let [top (first heap)
            moved (assoc (vec (butlast heap)) 0 (last heap))]
        [top (pq-sift-down moved 0)]))))

(def heap (reduce pq-push [] [5 2 8 1 9 3]))
(println "heap:" heap)

(loop [h heap acc []]
  (if (empty? h)
    (println "sorted via pops:" acc)
    (let [[top rest-h] (pq-pop h)]
      (recur rest-h (conj acc top)))))
